### DISCUSSION TITLE
In App class, self.tree does not contain all screens

### DISCUSSION THREAD

**Discussion by chnoma** [Textualize/textual#2295] [Q&A] [answered]
As the title says, when using `push_screen`, `self.tree` is changed in a way that prevents modifying elements on the previous screen. `self.query` for a previously existing object will throw an exception.

Here is a minimal example:
```
from datetime import datetime
from textual.app import App, ComposeResult
from textual.screen import ModalScreen
from textual.widgets import Static, Button, Label
from textual.containers import Grid


class QuitScreen(ModalScreen):
    """Screen with a dialog to quit."""

    def compose(self) -> ComposeResult:
        yield Grid(
            Label("Are you sure you would like to quit?", id="question"),
            Button("Quit", variant="error", id="quit"),
            Button("Cancel", variant="primary", id="cancel"),
            id="dialog",
        )

    def on_button_pressed(self, event: Button.Pressed) -> None:
        if event.button.id == "quit":
            self.app.exit()
        else:
            self.app.pop_screen()


class Clock(App):
    BINDINGS = [("x", "request_exit", "Exit Clock")]

    async def on_load(self) -> None:
        self.set_interval(1, self.update_loop)

    def compose(self) -> ComposeResult:
        yield Static("time", id="time")

    def update_loop(self) -> None:
        time_label = self.query_one("#time")
        time_label.update(str(datetime.now()))

    async def action_request_exit(self) -> None:
        print("BEFORE:")
        self.log(self.tree)
        qs = QuitScreen()
        await self.push_screen(qs)
        print("AFTER:")
        self.log(self.tree)

if __name__ == "__main__":
    app = Clock()
    app.run()
```

This will show the App's tree before and after `push_screen`, showing that the first screen with id `_default` is unavailable once the quit screen is pushed, and because of that, any children of `_default`:

```
[15:07:26] PRINT                                                                                                  query_fail_example.py:41
BEFORE:
[15:07:26] INFO                                                                                                   query_fail_example.py:42
Clock(title='Clock', classes={'-dark-mode'})
└── Screen(id='_default', pseudo_classes={'hover', 'enabled'})
    └── Static(id='time', pseudo_classes={'enabled'})
[15:07:26] SYSTEM                                                                                                              app.py:1454
QuitScreen(pseudo_classes={'enabled'}) is current (PUSHED)
[15:07:26] PRINT                                                                                                  query_fail_example.py:45
AFTER:
[15:07:26] INFO                                                                                                   query_fail_example.py:46
Clock(title='Clock', classes={'-dark-mode'})
└── QuitScreen(pseudo_classes={'enabled'})
    └── Grid(id='dialog', pseudo_classes={'enabled'})
        ├── Label(id='question', pseudo_classes={'enabled'})
        ├── Button(id='quit', classes={'-error'}, pseudo_classes={'enabled'}, variant='error')
        └── Button(id='cancel', classes={'-primary'}, pseudo_classes={'enabled'}, variant='primary')
```

Textual Diagnose:
# Textual Diagnostics

## Versions

| Name    | Value  |
|---------|--------|
| Textual | 0.19.1 |
| Rich    | 13.3.3 |

## Python

| Name           | Value                                                                     |
|----------------|---------------------------------------------------------------------------|
| Version        | 3.11.3                                                                    |
| Implementation | CPython                                                                   |
| Compiler       | MSC v.1934 64 bit (AMD64)                                                 |
| Executable     | C:\Users\Administrator\AppData\Local\Programs\Python\Python311\python.exe |

## Operating System

| Name    | Value      |
|---------|------------|
| System  | Windows    |
| Release | 10         |
| Version | 10.0.19042 |

## Terminal

| Name                 | Value            |
|----------------------|------------------|
| Terminal Application | Windows Terminal |
| TERM                 | *Not set*        |
| COLORTERM            | *Not set*        |
| FORCE_COLOR          | *Not set*        |
| NO_COLOR             | *Not set*        |

## Rich Console options

| Name           | Value                |
|----------------|----------------------|
| size           | width=138, height=34 |
| legacy_windows | False                |
| min_width      | 1                    |
| max_width      | 138                  |
| is_terminal    | True                 |
| encoding       | utf-8                |
| max_height     | 34                   |
| justify        | None                 |
| overflow       | None                 |
| no_wrap        | False                |
| highlight      | None                 |
| markup         | None                 |
| height         | None                 |

**Comment by davep** [comment]
> This will show the App's tree before and after push_screen, showing that the first screen with id _default is unavailable once the quit screen is pushed, and because of that, any children of _default

This would make sense. Once you push the quit screen it is the application's screen, and so the app's tree will be composed of that screen. 

The exception you get here from `query_one` is because you're doing the query on the app, which will mean that the query will then happen on `app.screen`, which at that point is the quit screen, not the default screen. That's what the `tree` is showing and that would be correct.

This feels like the sort of time where you don't want to be relying on using the default screen, and don't want to be doing the main work in the app; but instead should create a main screen that you're in control of and do the update work in there. For example:

```python
from datetime import datetime
from textual.app import App, ComposeResult
from textual.screen import ModalScreen, Screen
from textual.widgets import Static, Button, Label
from textual.containers import Grid


class QuitScreen(ModalScreen):
    """Screen with a dialog to quit."""

    def compose(self) -> ComposeResult:
        yield Grid(
            Label("Are you sure you would like to quit?", id="question"),
            Button("Quit", variant="error", id="quit"),
            Button("Cancel", variant="primary", id="cancel"),
            id="dialog",
        )

    def on_button_pressed(self, event: Button.Pressed) -> None:
        if event.button.id == "quit":
            self.app.exit()
        else:
            self.app.pop_screen()


class ClockScreen(Screen):

    def compose(self) -> ComposeResult:
        yield Static(str(datetime.now()), id="time")

    def on_mount(self) -> None:
        self.set_interval(1, self.update_loop)

    def update_loop(self) -> None:
        self.query_one("#time", Static).update(str(datetime.now()))


class Clock(App):
    BINDINGS = [("x", "request_exit", "Exit Clock")]

    def on_mount(self):
        self.push_screen(ClockScreen())

    def action_request_exit(self) -> None:
        self.push_screen(QuitScreen())

if __name__ == "__main__":
    app = Clock()
    app.run()
```

**Reply by chnoma** [reply]
Thank you! This makes a lot more sense -- as I mentioned in the discord I was conflating the Screen and App, but wasn't quite sure exactly how.

**Reply by davep** [reply]
Most welcome; and thank you for reducing the problem you were seeing down to the most basic example, it made it a lot easier to see what the issue was then.

**Comment by willmcgugan** [comment, marked as answer]
This is intentional. There is only every one screen in the DOM (the current screen), for the following reasons:

- It avoids making changes to widgets you didn't intend. 
- It's faster to query if you don't need to search all the screens.

If you want to get a widget from a specific screen you could pass a reference to the screen instance, or get it via the app with [get_screen](https://textual.textualize.io/api/app/#textual.app.App.get_screen).

**Reply by chnoma** [reply]
I see now, thank you very much!
